tests/test_facebook_attribution_window.py

```diff
@@ -0,0 +1,89 @@
+
+class FacebookAttributionWindow(FacebookBaseTest):
+
+    # set attribution window
+    ATTRIBUTION_WINDOW = 7
+
+    @staticmethod
+    def name():
+        return "tap_tester_facebook_attribution_window"
+
+    def get_properties(self, original: bool = True):
+        """Configuration properties required for the tap."""
+        return_value = {
+            'account_id': os.getenv('TAP_FACEBOOK_ACCOUNT_ID'),
+            'start_date' : '2019-07-24T00:00:00Z',
+            'end_date' : '2019-07-26T00:00:00Z',
+            'insights_buffer_days': str(self.ATTRIBUTION_WINDOW)
+        }
+        if original:
+            return return_value
+
+        return_value["start_date"] = self.start_date
+        return return_value
+
+    def test_run(self):
+        self.run_test(self.ATTRIBUTION_WINDOW) # attribution window: 7
+
+        self.ATTRIBUTION_WINDOW = 28
+        self.run_test(self.ATTRIBUTION_WINDOW) # attribution window: 28
+
+    def run_test(self, attr_window):
+        """
+            Test to check the attribution window
+        """
+
+        conn_id = connections.ensure_connection(self)
+
+        # get start date
+        start_date = self.get_properties()['start_date']
+        # calculate start date with attribution window
+        start_date_with_attribution_window = self.timedelta_formatted(start_date, days=-attr_window)
+
+        # 'attribution window' is only supported for 'ads_insights' streams
+        expected_streams = []
+        for stream in self.expected_streams():
+            if self.is_insight(stream):
+                expected_streams.append(stream)
+
+        # Run in check mode
+        found_catalogs = self.run_and_verify_check_mode(conn_id)
+
+        # Select only the expected streams tables
+        catalog_entries = [ce for ce in found_catalogs if ce['tap_stream_id'] in expected_streams]
+        self.perform_and_verify_table_and_field_selection(conn_id, catalog_entries, select_all_fields=True)
+
+        # Run a sync job using orchestrator
+        self.run_and_verify_sync(conn_id)
+        sync_records = runner.get_records_from_target_output()
+
+        expected_replication_keys = self.expected_replication_keys()
+
+        for stream in expected_streams:
+            with self.subTest(stream=stream):
+
+                replication_key = next(iter(expected_replication_keys[stream]))
+
+                # get records
+                records = [record.get('data') for record in sync_records.get(stream).get('messages')]
+
+                # check for the record is between attribution date and start date
+                is_between = False
+
+                for record in records:
+                    replication_key_value = record.get(replication_key)
+
+                    # Verify the sync records respect the (simulated) start date value
+                    self.assertGreaterEqual(self.parse_date(replication_key_value), self.parse_date(start_date_with_attribution_window),
```

Comment:
@harshpatel4crest Can you please explain how does the 3 checks asked in the DoD of this ticket is accumulated in this test case?

Reply:
`self.assertGreaterEqual(self.parse_date(replication_key_value), self.parse_date(start_date_with_attribution_window)` is for the 1st check.
`self.assertTrue(is_between, msg="No record found between start date and attribution date.")` is for the 2nd check, where `is_between` is the boolean value for the checking if any record is present between the start date and buffered start date.
3rd check is about the message that will be displayed during assertion.

Reply:
Please complete the 3rd assertion. If the tap does not throw an acceptable error, that is a bug and should be documented and linked in the test with a workaround.

Reply:
Facebook only supports those values, so even though the tap does not blow up when given a different value I think we should force an error so that we match what they claim they support. Additionally we have use cases where the tap is ran directly so, just because the UI has restrictions around the values does not mean a user will not send something crazy into the config.

FB Docs: https://www.facebook.com/business/help/395050428485124?id=428636648170202

Stitch Docs: https://www.stitchdata.com/docs/integrations/saas/facebook-ads#select-attribution-window

Reply:
@kspeer825 Added code change to raise an error when the attribution window is not 1, 7 or 28.